File: packages/rust/kernel-daemon/tests/daemon.rs
use std::io::{BufRead, BufReader, Write};
use std::net::TcpStream;
use std::process::{Child, Command, Stdio};
use std::thread;
use std::time::Duration;

use serde_json::{json, Value};

struct Daemon {
    child: Child,
    address: String,
}

impl Daemon {
    fn start() -> Self {
        let mut child = Command::new(env!("CARGO_BIN_EXE_graphvideo-kernel-daemon"))
            .env("GRAPHVIDEO_DAEMON_TOKEN", "fixture-secret-0001")
            .stdout(Stdio::piped())
            .stderr(Stdio::piped())
            .spawn()
            .expect("start daemon");
        let mut line = String::new();
        BufReader::new(child.stdout.take().unwrap())
            .read_line(&mut line)
            .expect("read daemon address");
        let ready: Value = serde_json::from_str(&line).expect("daemon ready JSON");
        Self {
            child,
            address: ready["address"].as_str().unwrap().to_owned(),
        }
    }

    fn connect(&self) -> Client {
        let stream = TcpStream::connect(&self.address).expect("connect to daemon");
        stream
            .set_read_timeout(Some(Duration::from_secs(3)))
            .unwrap();
        Client {
            reader: BufReader::new(stream.try_clone().unwrap()),
            writer: stream,
            next_id: 0,
        }
    }
}

impl Drop for Daemon {
    fn drop(&mut self) {
        let _ = self.child.kill();
        let _ = self.child.wait();
    }
}

struct Client {
    reader: BufReader<TcpStream>,
    writer: TcpStream,
    next_id: u64,
}

impl Client {
    fn request(&mut self, op: &str, args: Value) -> Value {
        self.next_id += 1;
        let mut request =
            json!({"version":1,"id":self.next_id,"token":"fixture-secret-0001","op":op});
        request
            .as_object_mut()
            .unwrap()
            .extend(args.as_object().unwrap().clone());
        writeln!(self.writer, "{request}").unwrap();
        let mut line = String::new();
        self.reader.read_line(&mut line).expect("daemon response");
        let response: Value = serde_json::from_str(&line).expect("response JSON");
        assert_eq!(response["id"], self.next_id);
        response
    }

    fn call(&mut self, op: &str, args: Value) -> Value {
        let response = self.request(op, args);
        assert_eq!(response["ok"], true, "{response}");
        response["result"].clone()
    }
}

#[test]
fn unauthorized_and_invalid_changes_do_not_mutate_state() {
    let daemon = Daemon::start();
    let mut client = daemon.connect();
    let unauthorized = client.request(
        "admit",
        json!({
            "token":"wrong-secret","nodeId":"bad","initialState":{}
        }),
    );
    assert_eq!(unauthorized["ok"], false);
    client.call(
        "admit",
        json!({"nodeId":"owner","initialState":{"count":0}}),
    );
    client.call("claim", json!({"nodeIds":["owner"]}));
    let first = client.call(
        "inject",
        json!({
            "targetNodeId":"owner","info":{"type":"IncrementInfo"},"submissionId":"s1"
        }),
    );
    assert_eq!(first["duplicate"], false);
    let duplicate = client.call(
        "inject",
        json!({
            "targetNodeId":"owner","info":{"type":"IncrementInfo"},"submissionId":"s1"
        }),
    );
    assert_eq!(duplicate["duplicate"], true);
    let change = client.call("poll", json!({}));
    assert_eq!(change["change"]["nodeId"], "owner");
    let invalid = client.request(
        "commit",
        json!({
            "changeId":change["change"]["changeId"],"operations":[
                {"op":"write","key":"count","value":1},
                {"op":"send","targetNodeId":"owner","info":{"bad":true}}
            ]
        }),
    );
    assert_eq!(invalid["ok"], false);
    assert_eq!(
        client.call("projection", json!({}))["nodes"]["owner"]["state"]["count"],
        0
    );
    client.call(
        "commit",
        json!({"changeId":change["change"]["changeId"],"operations":[]}),
    );
    assert_eq!(client.call("poll", json!({})), Value::Null);
}

#[test]
fn state_intervention_uses_generation_and_version_preconditions() {
    let daemon = Daemon::start();
    let mut client = daemon.connect();
    client.call(
        "admit",
        json!({"nodeId":"owner","initialState":{"count":0}}),
    );
    let edited = client.call(
        "intervene",
        json!({
            "nodeId":"owner","patch":{"count":5},"expectedGeneration":0,"expectedVersion":0
        }),
    );
    assert_eq!(edited["version"], 1);
    assert_eq!(edited["state"]["count"], 5);
    let conflict = client.request(
        "intervene",
        json!({
            "nodeId":"owner","patch":{"count":9},"expectedGeneration":0,"expectedVersion":0
        }),
    );
    assert_eq!(conflict["ok"], false);
    assert_eq!(
        client.call("projection", json!({}))["nodes"]["owner"]["state"]["count"],
        5
    );
}

#[test]
fn node_claim_is_exclusive_and_released_when_the_worker_disconnects() {
    let daemon = Daemon::start();
    let mut owner = daemon.connect();
    owner.call("admit", json!({"nodeId":"worker.node","initialState":{}}));
    owner.call("claim", json!({"nodeIds":["worker.node"]}));

    let mut contender = daemon.connect();
    let conflict = contender.request("claim", json!({"nodeIds":["worker.node"]}));
    assert_eq!(conflict["ok"], false);
    drop(owner);
    let mut reclaimed = false;
    for _ in 0..20 {
        let response = contender.request("claim", json!({"nodeIds":["worker.node"]}));
        if response["ok"] == true {
            reclaimed = true;
            break;
        }
        thread::sleep(Duration::from_millis(5));
    }
    assert!(reclaimed, "worker lease was not released after disconnect");
    assert_eq!(contender.call("health", json!({}))["leases"], 1);
}

#[test]
fn long_poll_wakes_when_another_client_injects_work() {
    let daemon = Daemon::start();
    let mut control = daemon.connect();
    control.call("admit", json!({"nodeId":"worker.node","initialState":{}}));
    let mut worker = daemon.connect();
    worker.call("claim", json!({"nodeIds":["worker.node"]}));
    let waiting = thread::spawn(move || worker.call("poll", json!({"waitMs":1000})));
    thread::sleep(Duration::from_millis(25));
    control.call(
        "inject",
        json!({
            "targetNodeId":"worker.node","info":{"type":"RunInfo"},"submissionId":"wake/1"
        }),
    );
    let change = waiting.join().unwrap();
    assert_eq!(change["change"]["nodeId"], "worker.node");
}

#[test]
fn effect_capability_brokers_opaque_requests_without_business_semantics() {
    let daemon = Daemon::start();
    let mut control = daemon.connect();
    control.call(
        "admit",
        json!({
            "nodeId":"world","initialState":{"result":null},
            "effectCapabilities":["fixture/effect"]
        }),
    );
    control.call(
        "inject",
        json!({
            "targetNodeId":"world","info":{"type":"RunInfo"},"submissionId":"effect/1"
        }),
    );
    let mut worker = daemon.connect();
    worker.call("claim", json!({"nodeIds":["world"]}));
    let change = worker.call("poll", json!({}));

    let unavailable = worker.request("requestEffect", json!({
        "changeId":change["change"]["changeId"],"adapterId":"fixture/effect","request":{"value":4}
    }));
    assert_eq!(unavailable["ok"], false);

    let mut provider = daemon.connect();
    provider.call("claimEffects", json!({"adapterIds":["fixture/effect"]}));
    let requested = worker.call("requestEffect", json!({
        "changeId":change["change"]["changeId"],"adapterId":"fixture/effect","request":{"value":4}
    }));
    let effect = provider.call("pollEffect", json!({}));
    assert_eq!(effect["adapterId"], "fixture/effect");
    assert_eq!(effect["request"]["value"], 4);
    provider.call(
        "completeEffect",
        json!({
            "effectId":effect["effectId"],"ok":true,"observation":{"value":8}
        }),
    );
    let observed = worker.call("awaitEffect", json!({"effectId":requested["effectId"]}));
    assert_eq!(observed, json!({"ok":true,"value":{"value":8}}));
    worker.call(
        "commit",
        json!({
            "changeId":change["change"]["changeId"],"operations":[
                {"op":"write","key":"result","value":observed["value"]}
            ]
        }),
    );
    assert_eq!(
        control.call("projection", json!({}))["nodes"]["world"]["state"]["result"]["value"],
        8
    );
}

#[test]
fn disconnecting_mid_change_releases_single_flight_and_routes_an_error_info() {
    let daemon = Daemon::start();
    let mut worker = daemon.connect();
    worker.call("admit", json!({"nodeId":"owner","initialState":{}}));
    worker.call("admit", json!({"nodeId":"errors","initialState":{}}));
    worker.call("setErrorTarget", json!({"nodeId":"errors"}));
    worker.call("claim", json!({"nodeIds":["owner"]}));
    worker.call(
        "inject",
        json!({
            "targetNodeId":"owner","info":{"type":"RunInfo"},"submissionId":"s1"
        }),
    );
    let change = worker.call("poll", json!({}));
    assert_eq!(change["change"]["nodeId"], "owner");
    drop(worker);

    let mut recovery = daemon.connect();
    recovery.call("claim", json!({"nodeIds":["errors"]}));
    let error = recovery.call("poll", json!({"waitMs":1000}));
    assert_eq!(error["change"]["nodeId"], "errors");
    assert_eq!(error["change"]["info"]["type"], "@error/NodeFailed");
    assert_eq!(error["change"]["info"]["nodeId"], "owner");
    recovery.call(
        "commit",
        json!({"changeId":error["change"]["changeId"],"operations":[]}),
    );
    assert_eq!(
        recovery.call("projection", json!({}))["submissions"]["s1"]["status"],
        "completed"
    );
}

#[test]
fn state_and_causal_queue_survive_a_client_disconnect() {
    let daemon = Daemon::start();
    let pid = daemon.child.id();
    let mut first = daemon.connect();
    first.call(
        "admit",
        json!({"nodeId":"source","initialState":{"count":0}}),
    );
    first.call("admit", json!({"nodeId":"sink","initialState":{"seen":0}}));
    first.call("claim", json!({"nodeIds":["source"]}));
    first.call(
        "inject",
        json!({"targetNodeId":"source","info":{"type":"IncrementInfo"},"submissionId":"s1"}),
    );
    let source = first.call("poll", json!({}));
    assert_eq!(source["change"]["nodeId"], "source");
    assert_eq!(source["state"]["count"], 0);
    first.call(
        "commit",
        json!({"changeId":source["change"]["changeId"],"operations":[
            {"op":"write","key":"count","value":1},
            {"op":"send","targetNodeId":"sink","info":{"type":"CountChangedInfo","count":1}}
        ]}),
    );
    drop(first);

    let mut second = daemon.connect();
    assert_eq!(second.call("health", json!({}))["pid"], pid);
    second.call("claim", json!({"nodeIds":["sink"]}));
    let sink = second.call("poll", json!({}));
    assert_eq!(sink["change"]["nodeId"], "sink");
    assert_eq!(sink["change"]["info"]["count"], 1);
    second.call(
        "commit",
        json!({"changeId":sink["change"]["changeId"],"operations":[
            {"op":"write","key":"seen","value":1}
        ]}),
    );
    let projection = second.call("projection", json!({}));
    assert_eq!(projection["nodes"]["source"]["state"]["count"], 1);
    assert_eq!(projection["nodes"]["sink"]["state"]["seen"], 1);
    assert_eq!(projection["submissions"]["s1"]["status"], "completed");
}

File: packages/sdk/python/src/graphframework_sdk/daemon.py
"""Daemon JSON Lines client: the Python twin of the JS KernelDaemonClient."""

import asyncio
import json
from typing import Any

from graphframework_sdk.protocol import PROTOCOL_VERSION, ProtocolError


class KernelDaemonClient:
    """Thin DTO client; one connection, sequential ids, `id`-keyed replies."""

    def __init__(self, reader: asyncio.StreamReader, writer: asyncio.StreamWriter, token: str):
        self._reader = reader
        self._writer = writer
        self._token = token
        self._next_id = 0

    @staticmethod
    async def connect(address: str, token: str, timeout_s: float = 5.0) -> "KernelDaemonClient":
        if len(token) < 16:
            raise ValueError("Kernel daemon token must have at least 16 bytes")
        host, _, port_text = address.rpartition(":")
        host = host.strip("[]")
        port = int(port_text)
        reader, writer = await asyncio.wait_for(asyncio.open_connection(host, port), timeout_s)
        return KernelDaemonClient(reader, writer, token)

    async def request(self, op: str, payload: dict[str, Any] | None = None) -> Any:
        self._next_id += 1
        frame = {"version": PROTOCOL_VERSION, "id": self._next_id, "token": self._token, "op": op}
        frame.update(payload or {})
        self._writer.write((json.dumps(frame) + "\n").encode("utf-8"))
        await self._writer.drain()
        line = await self._reader.readline()
        if not line:
            raise ProtocolError("Kernel daemon connection closed")
        response = json.loads(line.decode("utf-8"))
        if response.get("id") != self._next_id:
            raise ProtocolError(f"Kernel daemon reply id mismatch: {response!r}")
        if response.get("ok") is not True:
            raise ProtocolError(str(response.get("error", "Kernel daemon request failed")))
        return response.get("result")

    async def health(self) -> Any:
        return await self.request("health")

    async def shutdown(self) -> Any:
        return await self.request("shutdown")

    async def admit(self, node_id: str, initial_state: dict[str, Any],
                   analysis_facts: Any = None,
                   effect_capabilities: list[str] | None = None) -> Any:
        return await self.request("admit", {
            "nodeId": node_id, "initialState": initial_state,
            "analysisFacts": analysis_facts, "effectCapabilities": effect_capabilities or [],
        })

    async def evict(self, node_id: str) -> Any:
        return await self.request("evict", {"nodeId": node_id})

    async def replace(self, node_id: str, initial_state: dict[str, Any], analysis_facts: Any = None,
                      effect_capabilities: list[str] | None = None) -> Any:
        return await self.request("replace", {
            "nodeId": node_id, "initialState": initial_state,
            "analysisFacts": analysis_facts, "effectCapabilities": effect_capabilities or [],
        })

    async def inject(self, target_node_id: str, info: dict[str, Any], submission_id: str) -> Any:
        return await self.request("inject", {
            "targetNodeId": target_node_id, "info": info, "submissionId": submission_id,
        })

    async def claim(self, node_ids: list[str]) -> Any:
        return await self.request("claim", {"nodeIds": node_ids})

    async def release(self, node_ids: list[str]) -> Any:
        return await self.request("release", {"nodeIds": node_ids})

    async def poll(self, wait_ms: int = 0) -> Any:
        return await self.request("poll", {"waitMs": wait_ms})

    async def commit(self, change_id: int, operations: list[dict[str, Any]], error: str | None = None) -> Any:
        payload: dict[str, Any] = {"changeId": change_id, "operations": operations}
        if error is not None:
            payload["error"] = error
        return await self.request("commit", payload)

    async def cancel(self, submission_id: str) -> Any:
        return await self.request("cancel", {"submissionId": submission_id})

    async def intervene(self, node_id: str, patch: dict[str, Any],
                       expected_generation: int, expected_version: int) -> Any:
        return await self.request("intervene", {
            "nodeId": node_id, "patch": patch,
            "expectedGeneration": expected_generation, "expectedVersion": expected_version,
        })

    async def projection(self) -> Any:
        return await self.request("projection")

    async def analysis_facts(self) -> Any:
        return await self.request("analysisFacts")

    async def set_error_target(self, node_id: str) -> Any:
        return await self.request("setErrorTarget", {"nodeId": node_id})

    async def analyze(self, request: dict[str, Any]) -> Any:
        return await self.request("analyze", {"request": request})

    async def set_analysis_context(self, frontend_links: Any = None,
                                   frontend_service_links: Any = None) -> Any:
        return await self.request("setAnalysisContext", {
            "frontendLinks": frontend_links or [], "frontendServiceLinks": frontend_service_links or [],
        })

    async def agent_inspect(self, after: int | None = None, limit: int | None = None) -> Any:
        payload: dict[str, Any] = {}
        if after is not None:
            payload["after"] = after
        if limit is not None:
            payload["limit"] = limit
        return await self.request("agentInspect", payload)

    async def agent_inject(self, actor: str, reason: str, submission_id: str,
                           target_node_id: str, info: dict[str, Any]) -> Any:
        return await self.request("agentInject", {
            "actor": actor, "reason": reason, "submissionId": submission_id,
            "targetNodeId": target_node_id, "info": info,
        })

    async def agent_intervene_state(self, actor: str, reason: str, node_id: str,
                                    patch: dict[str, Any],
                                    expected_generation: int, expected_version: int) -> Any:
        return await self.request("agentInterveneState", {
            "actor": actor, "reason": reason, "nodeId": node_id, "patch": patch,
            "expectedGeneration": expected_generation, "expectedVersion": expected_version,
        })

    async def claim_effects(self, adapter_ids: list[str]) -> Any:
        return await self.request("claimEffects", {"adapterIds": adapter_ids})

    async def release_effects(self, adapter_ids: list[str]) -> Any:
        return await self.request("releaseEffects", {"adapterIds": adapter_ids})

    async def poll_effect(self, wait_ms: int = 0) -> Any:
        return await self.request("pollEffect", {"waitMs": wait_ms})

    async def request_effect(self, change_id: int, adapter_id: str, request: Any) -> Any:
        return await self.request("requestEffect", {
            "changeId": change_id, "adapterId": adapter_id, "request": request,
        })

    async def await_effect(self, effect_id: int, wait_ms: int = 30_000) -> Any:
        return await self.request("awaitEffect", {"effectId": effect_id, "waitMs": wait_ms})

    async def complete_effect(self, effect_id: int, ok: bool,
                              observation: Any = None, error: str | None = None) -> Any:
        payload: dict[str, Any] = {"effectId": effect_id, "ok": ok}
        if observation is not None:
            payload["observation"] = observation
        if error is not None:
            payload["error"] = error
        return await self.request("completeEffect", payload)

    def close(self) -> None:
        self._writer.close()
